bitcoin/src/blockdata/script/mod.rs

```diff
@@ -201,24 +212,25 @@ pub fn read_uint(data: &[u8], size: usize) -> Result<usize, Error> {
 
 // We internally use implementation based on iterator so that it automatically advances as needed
 // Errors are same as above, just different type.
+#[cfg_attr(all(test, mutate), mutate)]
 fn read_uint_iter(data: &mut core::slice::Iter<'_, u8>, size: usize) -> Result<usize, UintError> {
+    // `size` represents the number of byes to decode, this will only ever be 4 or less for bitcoin
+    // since integers are 32 bit.
+    assert!(size <= 4);
+
     if data.len() < size {
-        Err(UintError::EarlyEndOfScript)
-    } else if size > usize::from(u16::max_value() / 8) {
-        // Casting to u32 would overflow
-        Err(UintError::NumericOverflow)
-    } else {
-        let mut ret = 0;
-        for (i, item) in data.take(size).enumerate() {
-            ret = usize::from(*item)
-                // Casting is safe because we checked above to not repeat the same check in a loop
-                .checked_shl((i * 8) as u32)
-                .ok_or(UintError::NumericOverflow)?
-                .checked_add(ret)
-                .ok_or(UintError::NumericOverflow)?;
-        }
-        Ok(ret)
+        return Err(UintError::EarlyEndOfScript);
     }
+
+    let mut ret = 0;
+    for (i, item) in data.take(size).enumerate() {
+        ret = usize::from(*item)
+            .overflowing_shl((i * 8) as u32)
+            .0 // Casting (and overflowing) is safe because i <= 3.
+            .checked_add(ret)
+            .ok_or(UintError::NumericOverflow)?;
```

Comment:
In  a1ddd6bb3f038512c3f7b0ce7e8b09cd65a08152:

Should this `ret =` be `ret +=`?

Reply:
Oh, no, there's a `checked_add` in the pile of things hanging off of it. Cool.